### ISSUE TITLE
Implement plugin framework (Trac #4)

### ISSUE THREAD

**Issue by cfobel** [wheeler-microfluidics/microdrop#8]
It would be great to have a plugin directory where the user could drop python scripts which are loaded dynamically at runtime, allowing for the integration of additional instrumentation (motors, sensors, etc.) into the experimental protocol.

[http://wehart.blogspot.com/2009/01/python-plugin-frameworks.html This blog post] describes some of the existing plugin frameworks available for python.

Migrated from http://microfluidics.utoronto.ca/ticket/4
```json
{
    "status": "closed", 
    "changetime": "2014-04-17T19:39:01", 
    "description": "It would be great to have a plugin directory where the user could drop python scripts which are loaded dynamically at runtime, allowing for the integration of additional instrumentation (motors, sensors, etc.) into the experimental protocol.\n\n[http://wehart.blogspot.com/2009/01/python-plugin-frameworks.html This blog post] describes some of the existing plugin frameworks available for python.", 
    "reporter": "ryan", 
    "cc": "", 
    "resolution": "fixed", 
    "_ts": "1397763541728826", 
    "component": "microdrop", 
    "summary": "Implement plugin framework", 
    "priority": "major", 
    "keywords": "plugins", 
    "version": "0.1", 
    "time": "2011-09-22T01:12:02", 
    "milestone": "Microdrop 1.0", 
    "owner": "ryan", 
    "type": "enhancement"
}
```

**Comment by cfobel** [comment]
Trac update at 20111005T04:59:21: 
* ryan changed description from:

> It would be great to have a plugin directory where the user can drop python scripts which are loaded dynamically at runtime, allowing for the integration of additional instrumentation (motors, sensors, etc.) into the experimental protocol.
> 
> This blog post describes some of the existing plugin frameworks available for python:
> 
> http://wehart.blogspot.com/2009/01/python-plugin-frameworks.html

to:

> It would be great to have a plugin directory where the user could drop python scripts which are loaded dynamically at runtime, allowing for the integration of additional instrumentation (motors, sensors, etc.) into the experimental protocol.
> 
> [http://wehart.blogspot.com/2009/01/python-plugin-frameworks.html This blog post] describes some of the existing plugin frameworks available for python.


* ryan changed summary from "Implement plugin infrastructure for the protocol" to "Implement plugin framework"

**Comment by cfobel** [comment]
Trac update at 20111006T19:11:05: ryan commented:

> Changeset 46c1c31a2bf3c3d953d35195667ccc1af7df3c04 provides an initial implementation based on the [https://software.sandia.gov/trac/pyutilib PyUtilib] [https://software.sandia.gov/trac/pyutilib/export/1831/pyutilib.component.doc/trunk/doc/plugin/pca.pdf Plugin Component architecture module]. Currently, any class defined in a *.py file within the "plugins" directory and implementing the IPlugin interface is automatically loaded and enabled.
> 
> The biggest remaining limitation is that the state of plugins is not yet stored when a protocol is saved.

**Comment by cfobel** [comment]
Trac update at 20111007T20:09:12: ryan@fobel.net commented:

> commit af9bf2a6db96b069ef62054a4ab09f66f2dd048a
> Author: Ryan Fobel <ryan@fobel.net>
> Date:   Fri Oct 7 15:45:45 2011 -0400
> 
>     Add !PluginManager class and IPlugin interface
>     
>     This file should have been included in commit 46c1c31a2bf3c3d953d35195667ccc1af7df3c04
>     
>     References #8

**Comment by cfobel** [comment]
Trac update at 20111008T04:11:42: 
* ryan@fobel.net commented:

> commit 8f2b7e1bbd7fff3976c757988bcb401827a0587b
> Author: Ryan Fobel <ryan@fobel.net>
> Date:   Fri Oct 7 23:43:13 2011 -0400
> 
>     Support saving state for plugins
>     
>     Add 2 new functions to the IPlugin interface: on_protocol_load
>     and on_protocol_save. When on_protocol_save() is called, plugins
>     can pickle any state information to a string and add it to the
>     plugin_data dictionary stored by the Protocol object. When a
>     protocol is loaded, the on_protocol_load() function is called for
>     each plugin with the version number and pickled state as
>     parameters.
>     
>     Closes #8.


* ryan@fobel.net changed resolution from "" to "fixed"
* ryan@fobel.net changed status from "new" to "closed"

**Comment by cfobel** [comment]
Trac update at 20140417T19:39:01: ryan changed component from "core" to "microdrop"
